keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,611 @@
+- Add related e2e and unit tests to ensure functionality and make sure there is no regression in
+existing behavior when runtime class is not specified in the CRI calls to container runtime.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Work on feedback and add additional tests as needed
+- Follow up on container runtime changes for containerd and CRI-O
+
+#### GA
+
+- Decision on GA will be made based on beta feedback
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: RuntimeClassInCRICalls
+  - Components depending on the feature gate: kubelet
+- [x] Other
+  - Describe the mechanism: `N/A`
+  - Will enabling / disabling the feature require downtime of the control
+    plane? `No`
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? `No`
+
+###### Does enabling the feature change any default behavior?
+
+No. Aim of this KEP to ensure default behavior is not changed. Any changes to CRI and kubelet will be under feature gate as mentioned in earlier sections.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes. All the code changes needed will be under feature gate `RuntimeClassInCriCalls` that will be off by default for alpha.
+Default exisiting behavior of kubelet will be ensured.
```

Comment:
If I enable this, and then I end up pulling the same image name/tag but for different runtime classes to a node, then I disable this feature, will kubelet get confused about what image to use for a new container with that image name/tag?

Similarly, what happens to running containers? I guess they will be fine, even if they restart? But if the the pod is rescheduled to a node with flag off then of course it would get the host-matched image, but that seems fine.

Ideally, we should be able to disable this without any extra clean up needed. But if we do need some cleanup, we need to document it here.

Reply:
If the different image versions of the same image exist, turning off feature-gate without cleanup step will cause problems. This would require a cleanup step. Adding in these details as well.